opentaskspal/src/main/java/org/dmfs/opentaskspal/readdata/PercentComplete.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2017 dmfs GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.dmfs.opentaskspal.readdata;
+
+import org.dmfs.android.contentpal.RowDataSnapshot;
+import org.dmfs.iterators.Function;
+import org.dmfs.optional.decorators.DelegatingOptional;
+import org.dmfs.optional.decorators.Mapped;
+import org.dmfs.tasks.contract.TaskContract.Tasks;
+
+
+/**
+ * @author Gabor Keszthelyi
+ */
+public final class PercentComplete extends DelegatingOptional<Integer>
+{
+    public PercentComplete(RowDataSnapshot<Tasks> task)
+    {
+        super(new Mapped<>(new Function<CharSequence, Integer>()
+        {
+            @Override
+            public Integer apply(CharSequence charSequence)
+            {
+                // TODO Extract these kind of functions?
+                return Integer.valueOf(charSequence.toString());
```

Comment:
Here is another question: does it make sense to add another method to `RowSnapshotData` called `numberData(…)` returning an `Optional<Number>`? This could make it easier in many cases.

Reply:
I see now that `Number` has these `intValue()`, `longValue()`, etc, methods that we could use in the client code, but how could you implement returning those `Number` values in `RowDataSnapshot` correctly with general code? Not knowing which `Number` subtype to use.

Reply:
my idea was to extend `RowSnapshotData` with a method like this
```java
@NonNull
Optional<Number> numberData(@NonNull String key);
```

A number `numberData` could then return an `Optional` like so
```java
public Optional<Number> numberData(@NonNull String key)
{
   return new Mapped<CharSequence, Number>(chars -> new StringNumber(chars.toString()), charData(key));
}

…

public final static class StringNumber extends Number
{
   // for some weird reason Integer.parseInt(…) et al. expect a String when they would work just as well with a CharSequence :-|

    private final String mData;
    public StringNumber(String data) {mData = data;}

    public int intValue()
    {
       return Integer.parseInt(mData);
    }

   public float floatValue()
   {
      return Float.parseFloat(mData);
   }


   //etc …
}
```

So this could almost count as a convenience method but I think it can really improve our code.

Reply:
Sounds nice to me. The type may need to be `T extends Number` I think, to be able to get an `Optional<Integer>` for example and use the the default value method like `.value(3)`.

Create a ticket in ContentPal?

Reply:
oh, that `Optional<Integer>` wouldn't work of course, since it's actually an `Optional<StringNumber>...   so I am not sure how to use a default value easily with `Optional<Number>`